src/services/codeFixProvider.ts

```diff
@@ -3,6 +3,14 @@ namespace ts {
     export interface CodeFix {
         errorCodes: number[];
         getCodeActions(context: CodeFixContext): CodeAction[] | undefined;
+        createCodeFixDiagnosticIfApplicable?(node: Node, context: CodeFixDiagnoseContext): Diagnostic | undefined;

```

Comment:
Ideally we want this to be syntactic in nature, passing in the program allows for getting the checker. this can be expensive if the codeFix author do not know what they are doing, consider not passing in the program, but just a sourceFile and options.